[jinyang628] iP

jinyang628 opened the pull request:
### DukePro

> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. It's

- text-based
- easy to learn
- ~FAST~ _SUPER FAST_ to use

All you need to do is,

1. download it from [here](https://www.youtube.com/watch?v=dQw4w9WgXcQ).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
 
<hr> 

If you are a Java programmer, you can use it to practice Java too. Here's the main method: 

```java
public static void main(String[] args) {
    new Duke("data/tasks.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

jinyang628 committed:
Level 6 done

jinyang628 committed:
Removed unused imports

jinyang628 committed:
Added Java Docs

jinyang628 committed:
Save tasks local hard disk

jinyang628 committed:
Merge branch 'branch-Level-7'

jinyang628 committed:
Implement date recognition

jinyang628 committed:
Merge branch 'branch-Level-8'

jinyang628 committed:
Refactor code with more OOP

jinyang628 committed:
Divide classes into packages

jinyang628 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

jinyang628 committed:
Automate project builds using Gradle

jinyang628 committed:
Add JUnit tests

jinyang628 committed:
Package the App as a JAR file

jinyang628 committed:
Add java docs to half the non-private methods

jinyang628 committed:
Tweak the code to comply with coding standard

jinyang628 committed:
Give users a way ton find a task by searching for a keyword

jinyang628 committed:
Merge branch 'branch-Level-9'

jinyang628 committed:
Apply Java Docs changes

jinyang628 committed:
Apply Java Docs changes

jinyang628 committed:
Merge branch 'branch-A-CodingStandard'

jinyang628 committed:
All changes rectified

jinyang628 committed:
Correct errors regarding testing

jinyang628 committed:
Add more Javadocs

jinyang628 committed:
Implement checkstyle to detect coding style violations

jinyang628 committed:
Merge branch 'A-CheckStyle'

jinyang628 committed:
Replace Text UI with JFX GUI

jinyang628 committed:
Merge branch 'branchLevel10'

leezhanpeng commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,37 @@
+        this.by = LocalDate.parse(by);
+    }
+
+    /**
+     * Returns a string description which contains the deadline of the task.
+     * @return a string representation of the deadline.
+     */
```
For @return tags in Javadoc, it is actually indicated to be optional for members like toString() as they are [self-explanatory](https://google.github.io/styleguide/javaguide.html#s7.3.1-javadoc-exception-self-explanatory), so there is no problem in giving one-liners like this:

> /**
> &ensp;* Returns a string description which contains the deadline of the task.
> &ensp;*/

Same for your other files.

leezhanpeng commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,74 @@
+        this.parser = new Parser();
+    }
+
+    /**
+     * You should have your own function to generate a response to user input.
+     * Replace this stub with your completed method.
+     */
```
May be good to modify this already?

Nixx162 replied:
This comment can be removed

leezhanpeng commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    public String listTasks() {
+        return this.ui.listMessage(tasks);
+    }
+
+    public String markTaskAsDone(int taskNumber) {
```
Really good naming of all your functions in your PR throughout, not just in this file.

leezhanpeng commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,41 @@
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (from: " + this.from.format(DateTimeFormatter.ofPattern("MMM d yyyy"))
+                + " to: " + this.to.format(DateTimeFormatter.ofPattern("MMM d yyyy")) + ")";
```
Good use of linebreak! A lot of the other developers simply write the full line without consideration of the readability.

leezhanpeng reviewed (commented):
Overall amazing PR, nicely done.

Nixx162 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,74 @@
+    public String getResponse(String input) {
+        try {
+            String actionWord = this.parser.parseActionWord(input);
+            if (actionWord.equals("bye")) {
+                this.taskList.save(this.storage);
+                Platform.exit();
+            } else if (actionWord.equals("list")) {
+                return this.taskList.listTasks();
+            } else if (actionWord.equals("mark")) {
+                int taskNumber = this.parser.parseTaskNumber(input);
+                return this.taskList.markTaskAsDone(taskNumber);
+            } else if (actionWord.equals("unmark")) {
+                int taskNumber = this.parser.parseTaskNumber(input);
+                return this.taskList.unmarkTask(taskNumber);
+            } else if (actionWord.equals("delete")) {
+                int taskNumber = this.parser.parseTaskNumber(input);
+                return this.taskList.deleteTask(taskNumber);
+            } else if (actionWord.equals("find")) {
+                String description = this.parser.parseFindDescription(input);
+                return this.taskList.filterTasks(description);
+            } else if (Arrays.stream(this.specialTasksKeywords).anyMatch(
+                    keyword -> keyword.toString().toLowerCase().equals(actionWord))) {
+                try {
+                    Task task = this.parser.parseAddTaskInput(input, actionWord);
+                    return this.taskList.addTask(task);
+                } catch (InvalidTaskException e) {
+                    System.out.println(e.getMessage());
+                }
+            } else {
+                throw new InvalidInputException("ERROR: Invalid input");
+            }
```
A switch-case control flow would probably be less wordier

Nixx162 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,41 @@
+     * Returns a string description which contains the starting and ending time of the event.
+     * @return a string representation of the event.
+     */
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (from: " + this.from.format(DateTimeFormatter.ofPattern("MMM d yyyy"))
+                + " to: " + this.to.format(DateTimeFormatter.ofPattern("MMM d yyyy")) + ")";
+    }
+
+    /**
+     * Returns the information associated with the event to be stored on local hard disk.
+     * @return the string representation of the event saved onto the local hard disk.
+     */
+    public String saveTask() {
+        return "E|" + (this.isDone() ? "X|" : " |") + this.getDescription() + "|" + this.from + "|" + this.to;
+    }
```
You can try using the String.format() method for string-building cases like this

Nixx162 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,65 @@
+     */
+    @Override
+    public String toString() {
+        String prefix = "";
```
There shouldn't be any need to initialise a value for the prefix variable

Nixx162 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,56 @@
+    public String listFoundTasks(List<Task> filteredTasks) {
+        StringBuilder sb = new StringBuilder();
+        sb.append("Here are the matching tasks in your list:\n");
+        for (int i = 0; i < filteredTasks.size(); i++) {
+            sb.append(String.format("%d.%s\n", i + 1, filteredTasks.get(i).toString()));
+        }
```
It probably would be good for you to have a custom toString() method for your TaskList class, so you can abstract away menial tasks like this

Nixx162 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,56 @@
+    public String listMessage(ArrayList<Task> tasks) {
+        StringBuilder sb = new StringBuilder();
+        sb.append("Here are the tasks in your list:\n");
+        for (int i = 0; i < tasks.size(); i++) {
+            Task task = tasks.get(i);
+            sb.append(String.format("%d.%s%n", i + 1, task));
+        }
```
Same comment about having a custom toString() method for your TaskList class

Nixx162 reviewed (commented):
Couple of nits about some syntactic choice and OOP structure but quite good overall. Regarding the directory structure, I think it's better to group related classes to sub-packages of the duke package, instead of just dumping it all down there

jinyang628 committed:
Generate a new JAR file

jinyang628 committed:
Implement assertion statements

There are no assertion statements throughout the program.

Assertion statements are important in ensuring that certain key values
are as expected to ensure that Duke runs properly.

Let's write assertion statements in the Event and Deadline classes so
that we can ensure that the important variables get initialised and stored
properly.

jinyang628 committed:
Merge pull request #1 from jinyang628/branch-A-Assertions

Implement assertion statements

jinyang628 committed:
Correct JavaDocs

jinyang628 committed:
Improve code quality with proper logical structuring

Code is not segregated into logical paragraphs.

If code is not segregated properly, it decreases readability as users
are not able to immediately understand the flow of the program.

Leave empty spaces between each logical block.

These empty spaces demarcate logical blocks.

jinyang628 committed:
Merge pull request #2 from jinyang628/branch-A-CodeQuality

Improve code quality with proper logical structuring

jinyang628 committed:
Detect duplicates

jinyang628 committed:
Merge branch 'branch-Extension'

jinyang628 committed:
Insert a UI.png file

jinyang628 committed:
Append User Guide

jinyang628 committed:
Merge branch 'branch-A-UserGuide'

jinyang628 committed:
Remove undesirable styling for User Guide

jinyang628 committed:
Remove bugs in find method

jinyang628 committed:
Fix bugs related to Parser
